Library/Homebrew/metaformula.rb

```diff
@@ -0,0 +1,68 @@
+require "formula"
+
+# A meta-formula is a special kind of formula which will
+# * Allow bulk management of several related formulae (only `install` now)
+# * extend the semantics of 'dependency' to casks
+#
+# As a result, {MetaFormula} differs from {Formula} in that
+# * `url` is not mandatory
+# * `version` is mandatory, (since it can not be deduced from `url` now)
+# * the keg directory can be empty now (which will influence the "empty installation" check in audit.rb)
+# * only :stable is preserved, :head and :devel are not used
+# * `depends_on` now accepts :cask, e.g.
+#     depends_on :cask => "cask_name"
+#
+
+class MetaFormula < Formula
+  # Allow specs with no url
+  def spec_eval(name)
+    super
+
+    return unless name.equal? :stable
+    # we allow :stable spec with no url here
+    # unnecessary to keep three specs
+    # if we do, {Formula#installed_prefix} will report an error when `brew install xxx`
+    @stable = self.class.send(:stable)
+    @stable.owner = self
+    # `Formula#determine_active_spec` should now raise no "formulae require at least a URL" error
+  end
+
+  # Ignore url error in attribute validation
+  def validate_attributes!
+    super
+  rescue FormulaValidationError => e
+    raise unless e.attr.equal? :url
+  end
+
+  # Metaformula is not designed to be bottled for now
+  def bottle_disabled?
+    true
+  end
+
+  class << self
+    # DSL not designed to be used in MetaFormula for now are disabled to avoid undefined behaviors
+    # This is subject to change if certain functions should be assigned to them.
+
+    # {.stable}, {.devel} and {.head} only serve as attribute readers in {MetaFormula}
```

Comment:
Should `devel` and `head` not maybe be completely disabled for MetaFormulae?

Reply:
These methods are called in the initializer of Formula
```ruby
```

Reply:
These methods are called as soon as in the initializer of Formula :
```ruby
def initializer
    ...
    spec_eval :stable
    spec_eval :devel
    spec_eval :head
end
```
And the return value is supposed to have `{.url}` and thus not to be `nil` (and thus probably can not be disabled)
```ruby
def spec_eval(name)
	spec = self.class.send(name)
	return unless spec.url
end
```
So maybe best methods I can come up with is to return a default value and prohibit others from passing a block to them, I think?

Reply:
It may be worth instead having a custom initialiser?